src/CHIPGraph.cc

```diff
@@ -184,24 +204,69 @@ void CHIPGraph::removeNode(CHIPGraphNode *Node) {
   }
 }
 
-void CHIPGraphExec::launch(CHIPQueue *Queue) {
-  logDebug("{} CHIPGraphExec::launch({})", (void *)this, (void *)Queue);
-  compile();
-  auto ExecQueueCopy = ExecQueues_;
-  while (ExecQueueCopy.size()) {
-    auto Nodes = ExecQueueCopy.front();
-    std::string NodesInThisLevel = "";
-    for (auto Node : Nodes) {
-      NodesInThisLevel += Node->Msg + " ";
+bool CHIPGraphExec::tryLaunchNative(CHIPQueue *Queue) {
+  bool UsedNativeGraph = false;
+  if (NativeGraph) {
+    if (NativeGraph->isFinalized()) {
+      logDebug("NativeGraph: launching existing graph");
+    } else {
+      logDebug("NativeGraph: constructed but failed to finalize");
+      return false;
     }
-    logDebug("Executing nodes: {}", NodesInThisLevel);
-    for (auto Node : Nodes) {
-      logDebug("Executing {}", Node->Msg);
-      Node->execute(Queue);
-      Queue->finish();
+  } else {
+    logDebug("NativeGraph: trying to construct");
+    NativeGraph.reset(Queue->createNativeGraph());
+    if (!NativeGraph)
+      return false;
+
+    for (auto &Node : OriginalGraph_->getNodes()) {
+      if (!NativeGraph->addNode(Node)) {
+        logError("NativeGraph: failed to add node of type: {}",
+                 Node->getType());
+        return false;
+      }
     }
 
-    ExecQueueCopy.pop();
+    if (!NativeGraph->finalize()) {
+      logDebug("NativeGraph: failed to finalize");
+      return false;
+    }
+  }
+
+  assert(NativeGraph->isFinalized());
+  if (Queue->enqueueNativeGraph(NativeGraph.get())) {
+    logDebug("NativeGraph: launched");
+    Queue->finish();
+    return true;
+  } else {
+    return false;
+  }
```

Comment:
Style nit: [no else after return](https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return):
```suggestion
  if (Queue->enqueueNativeGraph(NativeGraph.get())) {
    logDebug("NativeGraph: launched");
    Queue->finish();
    return true;
  }
  return false;
```